src/main/java/org/zalando/nakadi/repository/TopicRepositoryHolder.java

```diff
@@ -1,51 +1,41 @@
 package org.zalando.nakadi.repository;
 
-import com.codahale.metrics.MetricRegistry;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
 import org.springframework.stereotype.Component;
-import org.zalando.nakadi.config.NakadiSettings;
+import org.zalando.nakadi.domain.NakadiCursor;
+import org.zalando.nakadi.domain.PartitionStatistics;
 import org.zalando.nakadi.domain.Storage;
+import org.zalando.nakadi.domain.Timeline;
+import org.zalando.nakadi.exceptions.NakadiRuntimeException;
+import org.zalando.nakadi.exceptions.ServiceUnavailableException;
 import org.zalando.nakadi.exceptions.TopicRepositoryException;
-import org.zalando.nakadi.repository.kafka.KafkaFactory;
-import org.zalando.nakadi.repository.kafka.KafkaLocationManager;
-import org.zalando.nakadi.repository.kafka.KafkaSettings;
-import org.zalando.nakadi.repository.kafka.KafkaTopicRepository;
-import org.zalando.nakadi.repository.zookeeper.ZooKeeperHolder;
-import org.zalando.nakadi.repository.zookeeper.ZookeeperSettings;
-import org.zalando.nakadi.util.UUIDGenerator;
 
+import java.util.Collections;
 import java.util.HashMap;
+import java.util.List;
 import java.util.Map;
 import java.util.concurrent.locks.ReadWriteLock;
 import java.util.concurrent.locks.ReentrantReadWriteLock;
+import java.util.stream.Collectors;
 
 @Component
 public class TopicRepositoryHolder {
 
     private static final Logger LOG = LoggerFactory.getLogger(TopicRepositoryHolder.class);
-    private final NakadiSettings nakadiSettings;
-    private final KafkaSettings kafkaSettings;
-    private final ZookeeperSettings zookeeperSettings;
-    private final UUIDGenerator uuidGenerator;
-    private final MetricRegistry metricRegistry;
+
+    private final Map<Storage.Type, TopicRepositoryCreator> repositoryCreators;
     private final ReadWriteLock readWriteLock;
     private final Map<Storage, TopicRepository> storageTopicRepository;
 
     @Autowired
-    public TopicRepositoryHolder(final NakadiSettings nakadiSettings,
-                                 final KafkaSettings kafkaSettings,
-                                 final ZookeeperSettings zookeeperSettings,
-                                 final UUIDGenerator uuidGenerator,
-                                 final MetricRegistry metricRegistry) {
-        this.nakadiSettings = nakadiSettings;
-        this.kafkaSettings = kafkaSettings;
-        this.zookeeperSettings = zookeeperSettings;
-        this.uuidGenerator = uuidGenerator;
-        this.metricRegistry = metricRegistry;
+    public TopicRepositoryHolder(@Qualifier("kafka") final TopicRepositoryCreator kafkaRepository) {
         this.readWriteLock = new ReentrantReadWriteLock();
         this.storageTopicRepository = new HashMap<>();
+        this.repositoryCreators = new HashMap<>();
+        this.repositoryCreators.put(Storage.Type.KAFKA, kafkaRepository);
```

Comment:
I would prefer to see method getSupportedStorageType() in TopicRepositoryCreator.